File: examples/dump_reader/read.py
# lammpstools is now a package, you can directly import
# block_data and dump_reader from them.
from lammpstools import block_data
from lammpstools import dump_reader
from lammpstools import data_field   # for data_field_types
import sys

# Define a new dump reader for a LAMMPS dump file.
d = dump_reader.dump_reader( "../common_dump_files/polymer.dump",
                            file_format = "PLAIN", dump_format = "LAMMPS" )

dl = dump_reader.dump_reader( "bondinfo.dump.bin",
                              file_format = "BIN", dump_format = "LAMMPS",
                              is_local = True )
dl.set_column_headers( ['c_bonds[1]', 'c_bonds[2]', 'c_bonds[3]'] )
# We know all cols in bondinfo.dump are ints, so set the default type:
dl.set_default_column_type( "int" )
# Specify the headers for atom dump file up front: dump_reader
# guesses the type of some cols based on this.
d.set_column_headers( [ 'id', 'mol', 'type', 'x', 'y', 'z' ] )

c = 0

# for b in d:
#     if c > 0 and c%25 == 0:
#         print("At block",c,", t =", b.meta.t, " contains N =",
#               b.meta.N, " particles")
#         i = 1233
#         if b.meta.N > i:
#             print("Position of particle",i,"is",b.x[i])
#             print("Type is",b.types[i],"and mol is",b.mol[i])
#     c += 1

for b, bl in zip(d, dl):
    if c > 0 and c%25 == 0:
        print("At block",c,", t =", b.meta.t, " contains N =",
              b.meta.N, " particles and", bl.meta.N, "bonds")
        i = 1233
        if b.meta.N > i:
            print("Position of particle",i,"is",b.x[i])
            print("Type is",b.types[i],"and mol is",b.mol[i])
        if bl.meta.N > i:
            print("data of bond",i,"is [ ", end = "")
            for j in range(0, len(bl.data)):
                print( bl.data[j][i]," ", end = "")
            print(" ]")
            print("Grabbing by name works: ", end="")
            print( bl.data_by_name('c_bonds[1]')[i] )
            print( b.data_by_name('y')[i] )
    c += 1

File: examples/dump_reader/read.cpp
/*
  This C++ program is functionally identical to the Python
  script in terms of functionality. Use these to see how
  much slower (or faster) Python is.

  Compile as
  $CC -O3 -std=c++11 read.cpp -I../../cpp_lib -llammpstools -o read_cpp
*/

#include "block_data.hpp"
#include "block_data_access.hpp"
#include "dump_reader_lammps_plain.hpp"
#include "data_field.hpp"

int main( int argc, char **argv )
{
	using namespace lammps_tools;

	using readers::dump_reader_lammps_plain;
	using readers::dump_reader_lammps;

	dump_reader_lammps_plain dl( "bondinfo.dump", dump_reader_lammps::LOCAL );
	dump_reader_lammps_plain d ( "polymer.dump",  dump_reader_lammps::CUSTOM );

	block_data b, bl;
	std::vector<std::string> headers = {"id", "mol", "type", "x", "y", "z"};
	std::vector<int> types = { block_data::ID, block_data::MOL,
	                           block_data::TYPE, block_data::X,
	                           block_data::Y, block_data::Z };
	for( std::size_t i = 0; i < headers.size(); ++i ){
		d.set_column_header( i, headers[i], types[i] );
	}

	dl.default_col_type = data_field::INT;

	int c = 0;
	int status = d.next_block(b);
	status |= dl.next_block(bl);
	while( status == 0 ){
		if( c > 0 && c % 25 == 0 ){

			std::cout << "At block " << c << ", t = " << b.tstep
			          << " contains N = " << b.N << " particles and "
			          << bl.N << " bonds\n";
			int i = 1233;
			if( b.N > i ){
				const std::vector<double> &x = get_x( b );
				const std::vector<double> &y = get_y( b );
				const std::vector<double> &z = get_z( b );

				std::cout << "Position of particle " << i << " is [ "
				          << x[i] << " " << y[i] << " " << z[i]
				          << " ]\n";
			}
			if( bl.N > i ){
				std::cout << "data of bond " << i << " is [ ";
				for( std::size_t j = 0; j < bl.n_data_fields(); ++j ){
					const std::vector<int> &data =
						data_as<int>( &bl[j] );
					std::cout << data[i] << " ";
				}
				std::cout << "]\n";
			}
		}

		status = d.next_block(b);
		status |= dl.next_block(bl);
		++c;
	}
	return 0;
}
